File: week14/20/pr1_Integer_to_Roman.py
class Solution:
    def intToRoman(self, num: int) -> str:
        idx = 0
        ans = ""

        if num >= 1000: idx = 3
        elif num >= 100 : idx = 2
        elif num >= 10 : idx = 1
        else : idx = 0
        # num : 1000~3999
        if idx == 3:
            target = num//1000
            ans += "M"*target
            num -= 1000*target
            idx-=1
        # num : 999~100
        if idx == 2:
            target = num//100
            if target == 4:
                ans += "CD"
            elif target == 9:
                ans += "CM"
            else:
                if target<5:
                    ans += "C"*target
                else:
                    ans += "D"
                    ans += "C"*(target-5)
            num -= 100*target
            idx-=1
        # num : 99~10
        if idx == 1:
            target = num//10
            if target == 4:
                ans += "XL"
            elif target == 9:
                ans += "XC"
            else:
                if target<5:
                    ans += "X"*target
                else:
                    ans += "L"
                    ans += "X"*(target-5)
            num -= 10*target
            idx-=1
        # num : 9~1
        if idx == 0:
            target = num
            if target == 4:
                ans += "IV"
            elif target == 9:
                ans += "IX"
            else:
                if target<5:
                    ans += "I"*target
                else:
                    ans += "V"
                    ans += "I"*(target-5)
            num -= 10*target
            idx-=1

        return ans            

File: week14/20/pr1_Integer_to_Roman.cpp
#include <vector>
using namespace std;
#include <string>
#include <tuple>
#include <stdexcept>

class Solution {
public:
    string intToRoman(int num) {

        vector<tuple<string,int>> m = {
            {"M",1000},
            {"CM",900},
            {"D",500},
            {"CD",400},
            {"C",100},
            {"XC",90},
            {"L",50},
            {"XL",40},
            {"X",10},
            {"IX",9},
            {"V",5},
            {"IV",4},
            {"I",1}
        };
        string ans = "";
        for(const auto pair : m){
            while(num >= get<1>(pair)){
                ans += get<0>(pair);
                num -= get<1>(pair);
            }
        }
        return ans;
    }
};